### PR TITLE
Fix path normalization in resolve_from_cwd()

### LINKED ISSUE #2718: --allow-read and --allow-write whitelists broken
The issue:
```js
const passwd = Deno.readFileSync('./../../../../../../etc/passwd');
console.log(new TextDecoder().decode(passwd));
```
successfully reads `/etc/passwd` when run with `--allow-read=./` because [the whitelist check](https://github.com/denoland/deno/blob/master/cli/permissions.rs#L484) does not first [canonicalize](https://doc.rust-lang.org/std/path/struct.PathBuf.html#canonicalize.v) the path.

While I discovered this issue, it was @nayeem-rahman who pointed out the cause when I brought up the bug in [the gitter chat](https://gitter.im/denolife/Lobby). I figured I should make an issue for it so that it could be tracked properly, especially considering it's a fairly major issue (Untrusted code that is meant to have access to only one directory has access to the whole file system).

### PR DESCRIPTION
**Pull request by nayeemrmn**
Fixes #2718.

`test.js`:
```js
Deno.readDirSync("..");
console.log("OK");
```
`deno --no-prompt --allow-read="." test.js`
Before: `OK`
After: `error: Uncaught PermissionDenied: permission denied ...`

### CODE REVIEW

**Comment by ry** [comment]
Cool - but is there some way we can get a unit test for this?

**Comment by nayeemrmn** [comment]
@ry Done. I wanted to add something to `permissions.rs` as well, but the whitelist checker in there assumes paths that are already normalised (which is handled in individual ops via `resolve_from_cwd()`) so they'd still fail. This could do with a refactor.

File: cli/fs.rs [hunk 1]
```diff
@@ -139,20 +139,56 @@ pub fn resolve_from_cwd(path: &str) -> Result<(PathBuf, String), ErrBox> {
...
 }
+
+#[cfg(test)]
+mod tests {
+  use super::*;
+
+  #[test]
+  fn resolve_from_cwd_child() {
+    let cwd = std::env::current_dir().unwrap();
+    assert!(resolve_from_cwd("a").unwrap().0 == cwd.join("a"));
```
Comment (ry) [inline cli/fs.rs]: Use `assert_eq!`

File: cli/fs.rs [hunk 2]
```diff
@@ -139,20 +139,56 @@ pub fn resolve_from_cwd(path: &str) -> Result<(PathBuf, String), ErrBox> {
...
+  fn resolve_from_cwd_dot() {
+    let cwd = std::env::current_dir().unwrap();
+    assert!(resolve_from_cwd(".").unwrap().0 == cwd);
+  }
+
+  #[test]
+  fn resolve_from_cwd_parent() {
+    let cwd = std::env::current_dir().unwrap();
+    assert!(resolve_from_cwd("a/..").unwrap().0 == cwd);
+  }
```
Comment (ry) [inline cli/fs.rs]: When I copy these tests into master, all but one of them pass. Is that expected?
```
test fs::tests::resolve_from_cwd_dot ... ok
test fs::tests::resolve_from_cwd_absolute ... ok
test fs::tests::resolve_from_cwd_child ... ok
test fs::tests::resolve_from_cwd_parent ... FAILED
```

**Comment by ry** [comment]
Looks good - just a few comments --^

File: cli/fs.rs (hunk 2 above)
Comment (nayeemrmn) [reply cli/fs.rs]: Yep.

File: cli/fs.rs (hunk 1 above)
Comment (nayeemrmn) [reply cli/fs.rs]: Done 👍

**Comment by ChiKler** [comment]
This bug remains unfixed :(

**Comment by kitsonk** [comment]
@ChiKler 😕 I just ran it and it works as expected:

```shell
❯ deno run --allow-read="." test.js
error: Uncaught PermissionDenied: read access to "/Users/kitsonk/github", run again with the --allow-read flag
    at processResponse (core.js:224:13)
    at Object.jsonOpSync (core.js:248:12)
    at Object.readDirSync (deno:cli/rt/30_fs.js:109:12)
    at file:///Users/kitsonk/github/test/test.js:1:6
```

Why do you think it isn't fixed?

**Comment by ChiKler** [comment]
> @ChiKler 😕 I just ran it and it works as expected:
> 
> ```shell
> ❯ deno run --allow-read="." test.js
> error: Uncaught PermissionDenied: read access to "/Users/kitsonk/github", run again with the --allow-read flag
>     at processResponse (core.js:224:13)
>     at Object.jsonOpSync (core.js:248:12)
>     at Object.readDirSync (deno:cli/rt/30_fs.js:109:12)
>     at file:///Users/kitsonk/github/test/test.js:1:6
> ```
> 
> Why do you think it isn't fixed?

Here's an example (am I doing something wrong?):
https://github.com/ChiKler/-Deno-error-Uncaught-PermissionDenied-read-access-to-CWD-

**Comment by nayeemrmn** [comment]
@ChiKler I believe that's because you use `copySync()` from `std`, which (without a good reason) calls `path.resolve()` which requires CWD access. It shouldn't call `path.resolve()`, could you open an issue?

**Pull request merged**